File: lib/ranker.js
const async = require('async');
const cloudLocations = require('./cloud_locations.js');
const groundWetness = require('./ground_wetness.js');
const rainForecast = require('./rain_forecast.js');

function calcRank(wetness, forecast) {
    return (wetness + forecast) / 2;
}

function rank(highBound, lowBound, callback) {
    cloudLocations.getCloudLocations(highBound, lowBound, function(err, cloud_path) {
        if (err) {
            callback(err);
            return;
        }
        async.map(
            cloud_path,
            function(loc, callback) {
                var wetness = groundWetness.getGroundWetness(loc);
                rainForecast.getRainValue(
                    loc,
                    function(err, forecast) {
                        var rank = calcRank(wetness, forecast);
                        
                        callback(null, {'loc': loc, 'rank': rank});
                    }
                );
            }, function (err, output) {
                if (err) {
                    callback(err);
                } else {
                    callback(null, output);
                }
            }
        );
    });
}

module.exports = {
    rank: rank
};

File: ranker.py
def get_cloud_locations():
    return [(1, 1), (2, 1), (3, 1), (4, 1)]


def get_ground_wetness(loc):
    return 0.5


def get_forecast(loc):
    return 0.25


def calc_rank(wetness, forecast):
    return wetness * forecast


def main():
    cloud_path = get_cloud_locations()
    output = dict()
    for loc in cloud_path:
        wetness = get_ground_wetness(loc)
        forecast = get_forecast(loc)
        rank = calc_rank(wetness, forecast)
        output[loc] = rank
    
    print(output)

main()
